pkg/controller/component/component_version.go

```diff
@@ -100,61 +103,143 @@ func UpdateCompDefinitionImages4ServiceVersion(ctx context.Context, cli client.R
 	appsInDef := covertImagesFromCompDefinition(compDef)
-	appsByUser, err := findMatchedImagesFromCompVersions(compVersions, serviceVersion)
+	appsInVer, err := findMatchedImagesFromCompVersions(compVersions, serviceVersion)
 	if err != nil {
 		return err
 	}
 
-	apps := checkNMergeImages(serviceVersion, appsInDef, appsByUser)
+	apps := checkNMergeImages(serviceVersion, appsInDef, appsInVer)
 
-	checkNUpdateImage := func(c *corev1.Container) error {
-		var err error
-		app, ok := apps[c.Name]
-		switch {
-		case ok && app.err == nil:
-			c.Image = app.image
-		case ok:
-			err = app.err
-		default:
-			err = fmt.Errorf("no matched image found for container %s", c.Name)
+	if err = func() error {
+		checkNUpdateImage := func(c *corev1.Container) error {
+			var err error
+			app, ok := apps[c.Name]
+			switch {
+			case ok && app.err == nil:
+				c.Image = app.image
+			case ok:
+				err = app.err
+			default:
+				err = fmt.Errorf("no matched image found for container %s", c.Name)
+			}
+			return err
+		}
+		for i := range compDef.Spec.Runtime.InitContainers {
+			if err := checkNUpdateImage(&compDef.Spec.Runtime.InitContainers[i]); err != nil {
+				return err
+			}
+		}
+		for i := range compDef.Spec.Runtime.Containers {
+			if err := checkNUpdateImage(&compDef.Spec.Runtime.Containers[i]); err != nil {
+				return err
+			}
 		}
+		return nil
+	}(); err != nil {
 		return err
 	}
 
-	for i := range compDef.Spec.Runtime.InitContainers {
-		if err := checkNUpdateImage(&compDef.Spec.Runtime.InitContainers[i]); err != nil {
-			return err
-		}
-	}
-	for i := range compDef.Spec.Runtime.Containers {
-		if err := checkNUpdateImage(&compDef.Spec.Runtime.Containers[i]); err != nil {
-			return err
+	if err = func() error {
+		for name, action := range actionsToResolveImage(compDef) {
+			if action != nil && action.Exec != nil {
+				if app, ok := apps[name]; ok {
+					if app.err != nil {
+						return app.err
+					}
+					action.Exec.Image = app.image
+				}
+			}
 		}
+		return nil
+	}(); err != nil {
+		return err
 	}
+
 	return nil
 }
 
 func covertImagesFromCompDefinition(compDef *appsv1.ComponentDefinition) map[string]appNameVersionImage {
 	apps := make(map[string]appNameVersionImage)
-	checkNAdd := func(c *corev1.Container) {
-		if len(c.Image) > 0 {
-			apps[c.Name] = appNameVersionImage{
-				name:    c.Name,
-				version: compDef.Spec.ServiceVersion,
-				image:   c.Image,
+
+	func() {
+		checkNAdd := func(c *corev1.Container) {
+			if len(c.Image) > 0 {
+				apps[c.Name] = appNameVersionImage{
+					name:     c.Name,
+					version:  compDef.Spec.ServiceVersion,
+					image:    c.Image,
+					required: true,
+				}
 			}
 		}
+		for i := range compDef.Spec.Runtime.InitContainers {
+			checkNAdd(&compDef.Spec.Runtime.InitContainers[i])
+		}
+		for i := range compDef.Spec.Runtime.Containers {
+			checkNAdd(&compDef.Spec.Runtime.Containers[i])
+		}
+	}()
+
+	func() {
+		checkNAdd := func(name string, a *appsv1.Action) {
```

Comment:
this function doesn't "check" for something.

Reply:
Fixed